### PR TITLE
Attribute Base Access Control core functions

### PR DESCRIPTION
**Pull request by diegomasini**
We have been reworking PR #1513 with @andresgaragiola and @raullaprida to split it in smaller PRs. This is the first one and it includes the core functions needed for Attribute Base Access Control. 
## Description

Core functions for Attribute Based Access Control needed by Membership Services, the client and chaincode were factored out in a common package that can be included in any of the three components mentioned. Importing this package does not provide any dependency with the rest of the fabric, just with the utils and the primitives packages. These three packages with common functionality could be extracted in a separated library.
## Motivation and Context

This is a required feature to address #1139
## How Has This Been Tested?

Current behave tests has been run without errors.
Current unit tests has been run without errors.
Additional unit tests where included (attributes_test.go in core/crypto/abac).
A new behave test is pending we will work with @jeffgarratt on that.
## Checklist:

- [x] I have added a [Signed-off-by](https://github.com/hyperledger/fabric/blob/master/CONTRIBUTING.md#legal-stuff)
- [x] Either no new documentation is required by this change, OR I added new documentation
- [x] Either no new tests are required by this change, OR I added new tests
- [x] I have run [goimports](https://godoc.org/golang.org/x/tools/cmd/goimports), [go vet](https://golang.org/cmd/vet/), and [golint](https://github.com/golang/lint). I have cleaned up all valid errors and warnings in code I have added or modified. These tools may generate false positives. Don't be worried about ignoring some errors or warnings. The goal is clean, consistent, and readable code.

Signed-off-by: Diego Masini damasini@us.ibm.com / diego.masini@gmail.com

### CODE REVIEW

File: core/crypto/abac/attributes.go [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+	return key, err
+}
+
+//GetValueForAttribute derives the K for the attribute "attributeName" and returns the value
+func GetValueForAttribute(attributeName string, preK0 []byte, cert *x509.Certificate) ([]byte, error) {
+	_, value, err := getKAndValueForAttribute(attributeName, preK0, cert)
+	return value, err
+}
+
+func createABACHeaderEntry(preK0 []byte) (*pb.ABACMetadataEntry, error) {
```
Comment (srderson) [inline core/crypto/abac/attributes.go]: Should `error` be part of the return definition if the function never returns an error?

File: core/crypto/abac/attributes.go [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+	attKey := getAttributeKey(preK0, HeaderAttributeName)
+	return &pb.ABACMetadataEntry{HeaderAttributeName, attKey}, nil
+}
+
+func createABACMetadataEntry(attributeName string, preK0 []byte) (*pb.ABACMetadataEntry, error) {
```
Comment (srderson) [inline core/crypto/abac/attributes.go]: Should `error` be part of the return definition if the function never returns an error?

File: core/crypto/abac/attributes.go [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+		entry, err := createABACMetadataEntry(key, preK0)
+		if err == nil {
+			entries = append(entries, entry)
+		}
+	}
+	headerEntry, err := createABACHeaderEntry(preK0)
+	if err == nil {
+		entries = append(entries, headerEntry)
+	}
+	return &pb.ABACMetadata{metadata, entries}, nil
```
Comment (srderson) [inline core/crypto/abac/attributes.go]: Should any of the `err` variables be returned here?

**Comment by srderson** [comment]
Thanks @diegomasini, much easier to review a commit of this size. You probably want to squash your commits as there are so many for this change.

File: core/crypto/abac/attributes.go (hunk 1 above)
Comment (diegomasini) [inline core/crypto/abac/attributes.go]: Yes, the error comes from the getKAndValueForAttribute call, it could be an error or nil. Are you thinking about a different way to achieve this?

File: core/crypto/abac/attributes.go (hunk 2 above)
Comment (diegomasini) [inline core/crypto/abac/attributes.go]: Yes, the error comes from the getKAndValueForAttribute call, it could be an error or nil. Are you thinking about a different way to achieve this?

File: core/crypto/abac/attributes.go (hunk 1 above)
Comment (srderson) [inline core/crypto/abac/attributes.go]: My comment is for the `createABACHeaderEntry` function. It never calls `getKAndValueForAttribute`

File: core/crypto/abac/attributes.go (hunk 2 above)
Comment (srderson) [inline core/crypto/abac/attributes.go]: My comment is for the `createABACMetadataEntry` function. It never calls `getKAndValueForAttribute`

File: core/crypto/abac/attributes.go (hunk 3 above)
Comment (diegomasini) [inline core/crypto/abac/attributes.go]: You are right, let me fix it

**Comment by diegomasini** [comment]
@srderson I just removed the err variables from CreateABACMetadataObjectFromCert() you mentioned and I squashed the changes of this PR.

**Comment by srderson** [comment]
@adecaro please give this a quick review when you have a chance. Thanks

File: core/crypto/abac/attributes.go [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+var (
+	// TCertEncAttributesBase is the base ASN1 object identifier for attributes.
+	// When generating an extension to include the attribute an index will be
+	// appended to this Object Identifier.
+	TCertEncAttributesBase = asn1.ObjectIdentifier{1, 2, 3, 4, 5, 6}
+
+	// TCertAttributesHeaders is the ASN1 object identifier of attributes header.
+	TCertAttributesHeaders = asn1.ObjectIdentifier{1, 2, 3, 4, 5, 6, 9}
+
+	padding = []byte{255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255}
```
Comment (adecaro) [inline core/crypto/abac/attributes.go]: What is the purpose of padding and why does it have that specific value?

File: core/crypto/abac/attributes.go [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+	if err != nil {
+		return nil, encrypted, err
+	}
+	return value, encrypted, nil
+}
+
+//EncryptAttributeValue encrypts "attributeValue" using "attributeKey"
+func EncryptAttributeValue(attributeKey []byte, attributeValue []byte) ([]byte, error) {
+	value := append(attributeValue, padding...)
+	return primitives.CBCPKCS7Encrypt(attributeKey, value)
```
Comment (adecaro) [inline core/crypto/abac/attributes.go]: Because the attributes are used in multiple places, shall we use GCM mode to protect against malleability attacks? The primitives package offers such an encryption mode.

File: core/crypto/abac/attributes.go [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+}
+
+//getAttributeKey returns the attributeKey derived from the preK0 to the attributeName.
+func getAttributeKey(preK0 []byte, attributeName string) []byte {
+	return primitives.HMACTruncated(preK0, []byte(attributeName), 32)
```
Comment (adecaro) [inline core/crypto/abac/attributes.go]: Shall we add a prefix to []byte(attributeName) in order to distinguish the scope of the keys derived from preK0? 

Suppose that in the future we will use preK0 to derive keys wth different scopes. A way to identify the scope is by adding a prefix (associated to that specific scope) to the argument. This offers more modularity and retro compatibility. Of course, if we are pretty sure that preK0 will be used only for one thing, then we don't need this prefixing :)

File: core/crypto/abac/proto/abac.proto [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+
+package protos;
+
+//ABACMetadataEntry is an entry within the metadata that store an attribute name with its respective key.
+message ABACMetadataEntry {
+    string AttributeName = 1;
+    bytes AttributeKey = 2;
+}
+
+//ABACMetadata hold both the original metadata bytes and the metadata required by the Attribute Based Access Control
```
Comment (adecaro) [inline core/crypto/abac/proto/abac.proto]: Just a clarification: Is the marshaled version of ABACMetadata what we will have to put in the transaction metadata for access control?

File: core/crypto/abac/attributes.go [hunk 8]
```diff
@@ -0,0 +1,267 @@
...
+
+	return result, nil
+}
+
+//ReadAttributeHeader read the header of the attributes.
+func ReadAttributeHeader(tcert *x509.Certificate, headerKey []byte) (map[string]int, bool, error) {
+	var err error
+	var headerRaw []byte
+	encrypted := false
+	if headerRaw, err = utils.GetCriticalExtension(tcert, TCertAttributesHeaders); err != nil {
```
Comment (adecaro) [inline core/crypto/abac/attributes.go]: Just a clarification: Does the function assume that the x509 certificate has been already validated? I guess so, given that certificates must be always validated before extracting any information from them.

**Comment by adecaro** [comment]
Hi @diegomasini, a way to avoid to have all these "Merge remote-tracking branch 'upstream/master'" is to rebase rather then merging. This helps reducing the number of commits :)

File: core/crypto/abac/attributes.go (hunk 8 above)
Comment (diegomasini) [inline core/crypto/abac/attributes.go]: Yes, at this point the TCert is validated in the stub, that code is not part of this PR since we are splitting a bigger PR in smaller pieces.

File: core/crypto/abac/proto/abac.proto (hunk 7 above)
Comment (diegomasini) [inline core/crypto/abac/proto/abac.proto]: Yes, the use is part of the client. And there is also an example of its use but is part of a coming PR as part of the split of the initial PR.

File: core/crypto/abac/attributes.go (hunk 4 above)
Comment (diegomasini) [inline core/crypto/abac/attributes.go]: When a value is decrypted the decryption does not fail if someone uses the wrong key. It just produce garbage information. The padding was introduced to allow someone to verify that the decryption was successful. It needs to be a constant known value.

File: core/crypto/abac/attributes.go (hunk 6 above)
Comment (diegomasini) [inline core/crypto/abac/attributes.go]: Attributes are associated to a given affiliation. In order to request a batch of TCerts the client needs to provide its ECert. That way attributes are implicitly scoped to that affiliation. It is possible to add a prefix to the attribute name if needed.

File: core/crypto/abac/attributes.go (hunk 5 above)
Comment (diegomasini) [inline core/crypto/abac/attributes.go]: Attribute's keys are distributed in the transaction by using the provided confidentiality mechanism. I would like to review this with you since for the asset transfer use case we need to use GCM and potentially modify the transaction structure. But for the scope of this PR I think the confidentiality mechanism is enough. Also, I don't know is this is clear, but attributes are encrypted each time with a different key.

File: core/crypto/abac/attributes_test.go [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+
+func TestGetKAndValueForAttribute(t *testing.T) {
+	expected := "Software Engineer"
+
+	tcert, prek0, err := loadTCertAndPreK0()
+	if err != nil {
+		t.Error(err)
+	}
+
+	_, attribute, err := getKAndValueForAttribute("position", prek0, tcert)
```
Comment (christo4ferris) [inline core/crypto/abac/attributes_test.go]: why are we (only) testing private functions? Seems to me that we should be testing both, but most importantly the public functions. Also, we only have positive tests here. Should have tests that ensure the correct processing of calls that pass incorrect argument values.

File: core/crypto/abac/attributes_test.go [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+
+func TestGetValueForAttribute(t *testing.T) {
+	expected := "Software Engineer"
+
+	tcert, prek0, err := loadTCertAndPreK0()
+	if err != nil {
+		t.Error(err)
+	}
+
+	value, err := GetValueForAttribute("position", prek0, tcert)
```
Comment (christo4ferris) [inline core/crypto/abac/attributes_test.go]: should include cases for incorrect arguments and validate expected errors are returned, etc

File: core/crypto/abac/attributes_test.go [hunk 11]
```diff
@@ -0,0 +1,229 @@
...
+
+func TestReadTCertAttributeByPosition(t *testing.T) {
+	expected := "Software Engineer"
+
+	tcert, prek0, err := loadTCertAndPreK0()
+	if err != nil {
+		t.Error(err)
+	}
+
+	encryptedAttribute, err := ReadTCertAttributeByPosition(tcert, 1)
```
Comment (christo4ferris) [inline core/crypto/abac/attributes_test.go]: we should really be including more variability in the test cases. What if the function could only return the 1st position? please add more test cases.

File: core/crypto/abac/attributes_test.go [hunk 12]
```diff
@@ -0,0 +1,229 @@
...
+		t.Error("Empty header should produce a parsing error")
+	}
+}
+
+func TestBuildAndParseAttributesHeader(t *testing.T) {
+	attributes := make(map[string]int)
+	attributes["company"] = 1
+	attributes["position"] = 2
+
+	header := string(BuildAttributesHeader(attributes)[:])
```
Comment (christo4ferris) [inline core/crypto/abac/attributes_test.go]: this public function is only being indirectly tested, should have tests that check for things like passing an empty set of attributes, for instance.

File: core/crypto/abac/attributes.go [hunk 13]
```diff
@@ -0,0 +1,267 @@
...
+	cert, err := utils.DERToX509Certificate(raw)
+	if err != nil {
+		return nil, err
+	}
+
+	return CreateABACMetadataFromCert(cert, metadata, preK0, attributeKeys)
+}
+
+//GetABACMetadata object from the original metadata "metadata".
+func GetABACMetadata(metadata []byte) (*pb.ABACMetadata, error) {
```
Comment (christo4ferris) [inline core/crypto/abac/attributes.go]: Please add tests for this public function

File: core/crypto/abac/attributes.go [hunk 14]
```diff
@@ -0,0 +1,267 @@
...
+
+//CreateABACMetadataFromCert creates the ABACMetadata from the original metadata and certificate "cert".
+func CreateABACMetadataFromCert(cert *x509.Certificate, metadata []byte, preK0 []byte, attributeKeys []string) ([]byte, error) {
+	abacMetadata := CreateABACMetadataObjectFromCert(cert, metadata, preK0, attributeKeys)
+
+	return proto.Marshal(abacMetadata)
+}
+
+//CreateABACMetadata create the ABACMetadata from the original metadata
+func CreateABACMetadata(raw []byte, metadata []byte, preK0 []byte, attributeKeys []string) ([]byte, error) {
```
Comment (christo4ferris) [inline core/crypto/abac/attributes.go]: Please add tests for this

File: core/crypto/abac/attributes.go [hunk 15]
```diff
@@ -0,0 +1,267 @@
...
+		entry := createABACMetadataEntry(key, preK0)
+		entries = append(entries, entry)
+	}
+	headerEntry := createABACHeaderEntry(preK0)
+	entries = append(entries, headerEntry)
+
+	return &pb.ABACMetadata{metadata, entries}
+}
+
+//CreateABACMetadataFromCert creates the ABACMetadata from the original metadata and certificate "cert".
```
Comment (christo4ferris) [inline core/crypto/abac/attributes.go]: please add tests for this

File: core/crypto/abac/attributes.go [hunk 16]
```diff
@@ -0,0 +1,267 @@
...
+	return &pb.ABACMetadataEntry{HeaderAttributeName, attKey}
+}
+
+func createABACMetadataEntry(attributeName string, preK0 []byte) *pb.ABACMetadataEntry {
+	attKey := getAttributeKey(preK0, attributeName)
+	return &pb.ABACMetadataEntry{attributeName, attKey}
+}
+
+//CreateABACMetadataObjectFromCert creates an ABACMetadata object from certificate "cert", metadata and the attributes keys.
+func CreateABACMetadataObjectFromCert(cert *x509.Certificate, metadata []byte, preK0 []byte, attributeKeys []string) *pb.ABACMetadata {
```
Comment (christo4ferris) [inline core/crypto/abac/attributes.go]: please add tests for this

File: core/crypto/abac/attributes_test.go (hunk 9 above)
Comment (andresgaragiola) [inline core/crypto/abac/attributes_test.go]: TestGetKAndValueForAttribute is a test for private function "getKAndValueForAttribute", this private function is called by two public functions "GetValueForAttribute" and "GetKForAttribute". "GetValueForAttribute" is tested in the function TestGetValueForAttribute and we are adding a new test now called "TestGetKForAttribute" to test GetKForAttribute. Other public functions tests currently exists in this file. TestReadTCertAttributeByPosition_InvalidPositions is an example of a negative case test we will be adding more negative cases tests to this file.

**Comment by christo4ferris** [comment]
test coverage is currently 67% for this new component, please use a tool such as go-carpet to flesh out the remaining unit test coverage.

File: core/crypto/abac/attributes_test.go (hunk 11 above)
Comment (diegomasini) [inline core/crypto/abac/attributes_test.go]: I think I don't understand completely here, position is an attribute holding the position inside the company, not a position inside the TCert.

Comment (christo4ferris) [inline core/crypto/abac/attributes_test.go]: to clarify my comment: if the function only returned the cert at position 1, this test passes and everything looks fine... but requesting the cert at position 2 would fail - and clearly that would be a bug.

**Comment by diegomasini** [comment]
@srderson @binhn could you take a look at this PR again? We addressed the comments posted here and as @christo4ferris suggested we are using go-carpet to measure the testing coverage. With the last unit tests it is now 86.2% Remember that this PR needs to be processed to allow us to send the rest of the PRs that constitutes the whole ABAC feature.

**Comment by christo4ferris** [comment]
86% is decent thank you... I can't review in detail now but am pleased with the coverage.

**Comment by binhn** [comment]
@diegomasini Thanks Diego for your patience and collaboration on this pull request. 

I looked over this again tonight, and one last question: In `abac.proto`, you define `ABACMetadata`. Why metadata? The content obviously contains an array of attributes, not something that we don't know. Also `ABACMetadata` contains `Metadata` as bytes. It is not clear to me why we need both `Metadata` and `Entries` in this struct. Could you offer some explanation as part of comment of the struct.

**Comment by andresgaragiola** [comment]
thanks for your comment @binhn. ABACMetadata includes both the original metadata and one entry per each attribute included in the transaction. Each ABACMetadataEntry includes the attributeName and the key required to decrypt that attribute. We have added the metadata field to wrap the existing metadata as it is used currently. We also provided an API to read the attributes from it.

**Comment by binhn** [comment]
@diegomasini @andresgaragiola  As discussed, we agreed to change the package name, including files and functions, so that it would be more general since attributes are not only for abac but can be used anyhow the app wishes. We also agreed to defer the resolution to the problem of having the peer knows about the decryption key of the attribute keys, until we complete other parts of attributes so that other work may continue.

**Comment by andresgaragiola** [comment]
@binhn as we discussed today, we recently have pushed a new version of this library renamed to 'attributes' instead of 'abac' and some additional test increasing the coverage to 92%.

**Comment by binhn** [comment]
@andresgaragiola Excellent.  Thank you!
